src/ws/cockpitsshagent.c

```diff
@@ -0,0 +1,314 @@
+                                         agent);
+
+  options = json_object_new ();
+  json_object_set_string_member (options, "channel", agent->channel_id);
+  json_object_set_string_member (options, "command", "open");
+  json_object_set_string_member (options, "binary", "raw");
+  json_object_set_string_member (options, "payload", "stream");
+  json_object_set_string_member (options, "internal", "ssh-agent");
+  message = cockpit_json_write_bytes (options);
+  cockpit_transport_send (agent->transport,
+                          NULL,
+                          message);
+
+  agent->sig_pipe_read = g_signal_connect (agent->pipe,
+                                           "read",
+                                           G_CALLBACK (on_agent_pipe_read),
+                                           agent);
+  agent->sig_pipe_close = g_signal_connect (agent->pipe,
+                                            "close",
+                                            G_CALLBACK (on_agent_pipe_close),
+                                            agent);
+
+out:
+  if (options)
+    json_object_unref (options);
+  if (message)
+    g_bytes_unref (message);
+
+  return agent;
+}
+
+void
+cockpit_ssh_agent_close (CockpitSshAgent *agent)
+{
+  /* is agent open? */
+  if (!agent->open)
+    return;
+
+  g_debug ("%s: close agent", agent->logname);
+  agent->open = FALSE;
+
+  if (agent->sig_pipe_close > 0)
+    g_signal_handler_disconnect (agent->pipe, agent->sig_pipe_close);
+  agent->sig_pipe_close = 0;
+
+  if (agent->sig_pipe_read > 0)
+    g_signal_handler_disconnect (agent->pipe, agent->sig_pipe_read);
+  agent->sig_pipe_read = 0;
+
+  if (!agent->pipe_closed)
+    {
+      agent->pipe_closed = TRUE;
+      cockpit_pipe_close (agent->pipe, NULL);
+    }
+  if (agent->sig_closed > 0)
+    g_signal_handler_disconnect (agent->transport, agent->sig_closed);
+  agent->sig_closed = 0;
+
+  if (agent->sig_recv > 0)
+    g_signal_handler_disconnect (agent->transport, agent->sig_recv);
+  agent->sig_recv = 0;
+
+  if (agent->sig_control > 0)
+    g_signal_handler_disconnect (agent->transport, agent->sig_control);
+  agent->sig_control = 0;
+
+  if (!agent->transport_closed)
+    {
+      JsonObject *options = NULL;
+      GBytes *message = NULL;
+
+      agent->transport_closed = TRUE;
+      options = json_object_new ();
+      json_object_set_string_member (options, "channel", agent->channel_id);
+      json_object_set_string_member (options, "command", "close");
+      message = cockpit_json_write_bytes (options);
+      cockpit_transport_send (agent->transport,
+                              NULL,
+                              message);
+      json_object_unref (options);
+      g_bytes_unref (message);
+    }
+
+  g_object_unref (agent->pipe);
+  agent->pipe = NULL;
+
+  g_free (agent->channel_id);
+  agent->channel_id = NULL;
+}
+
+int
+cockpit_ssh_agent_get_fd (CockpitSshAgent *agent)
+{
+  return agent->ssh_fd;
```

Comment:
Where is ssh_fd closed? Does libssh close it? If so, we probably want to change how this function works. Such as cockpit_ssh_agent_claim_fd() and it shuold clear the ssh_fd and have checks against calling it twice, etc.